File: toyNN/NeuralNetwork.cpp
//
//  NeuralNetwork.cpp
//  toyNN
//
//  Created by Jeffrey Fisher on 2/11/20.
//  Copyright © 2020 Jeffrey Fisher. All rights reserved.
//

#include "NeuralNetwork.hpp"
#include <sstream>
#include <fstream>

NeuralNetwork::NeuralNetwork(){}

NeuralNetwork::NeuralNetwork(std::vector<int> numNodes, std::string fileName){
    this->fileName = fileName;
    this->numNodes = numNodes;
    for(int i=0; i < (this->numNodes.size()-1) ; i++){
        Layer newLayer(this->numNodes[i], this->numNodes[i+1]);
        this->layers.push_back(newLayer);
    }
    this->getWeightsFile();
}

NeuralNetwork::NeuralNetwork(std::string fileName){
    this->fileName = fileName;
    this->getWeightsFile();
}


NeuralNetwork::NeuralNetwork(const NeuralNetwork &obj){
    this->numNodes = obj.numNodes;
    for(int i=0; i < (this->numNodes.size()-1) ; i++){
        Layer newLayer(this->numNodes[i], this->numNodes[i+1]);
        this->layers.push_back(newLayer);
    }
}

NeuralNetwork::~NeuralNetwork(void){this->saveWeightsFile(this->fileName);}

void NeuralNetwork::setupNetworkFromFile(){
    std::ifstream savedFile;
    savedFile.open(this->fileName);
    if(savedFile.is_open()){
        int matrixInfo[4];
        std::string scannedValue;
        std::vector<int> numNodes;
        while(!savedFile.eof()){
            std::getline(savedFile, scannedValue);
            if(scannedValue == "~"){
                for(int j=0;j<4;j++){
                    std::getline(savedFile, scannedValue, ' ');
                    matrixInfo[j] = std::stoi(scannedValue);
                }
                if(matrixInfo[1]==0){
                    if(matrixInfo[0] == 0){
                        numNodes.push_back(matrixInfo[3]);
                    }
                    numNodes.push_back(matrixInfo[2]);
                }
            }
        }
        this->numNodes = numNodes;
        savedFile.close();
    } else {
        std::cout<<"ERROR: Could not open file."<<std::endl;
        exit(1);
    }
    for(int i=0; i < (this->numNodes.size()-1) ; i++){
        Layer newLayer(this->numNodes[i], this->numNodes[i+1]);
        this->layers.push_back(newLayer);
    }
}

Matrix NeuralNetwork::feedForward(std::vector<std::vector<double>> inputData){
    Matrix inputs(inputData);
    this->layers.front().setInputs(inputs, true);
    for(int j=1; j < this->layers.size(); j++){
        this->layers[j].setInputs(this->layers[j-1].solveForOutputs(), false);
    }
    return (this->layers[this->layers.size()-1].solveForOutputs());
}

void NeuralNetwork::train(std::vector<std::vector<double>> inputData, std::vector<std::vector<double>> targetData){
    if((inputData[0].size() != this->numNodes.front()) || (targetData[0].size() != this->numNodes.back())){
        std::cout<<"ERROR: Training data dimentions do not match network dimentions."<<std::endl;
        exit(1);
    } else {
        int x;
        for(int i=0; i < trainingRuns; i++){ //Repeat until model has ran for desired number of training runs
            x = rand() % (inputData.size()-1); //Select a random training set.
            
            std::vector<std::vector<double>> trainingInput; //2D vector with single input set
            trainingInput.push_back(inputData[x]);
            std::vector<std::vector<double>> trainingTarget; //2D vector with single output set
            trainingTarget.push_back(targetData[x]);
            
            Matrix trainingOutputs(this->feedForward(trainingInput)); //Matrix of outputs from the training run
            Matrix trainingTargets(trainingTarget);  //Matrix of targets from the training run
            trainingTargets.transpose();
            
            this->layers[this->layers.size()-1].outputError = trainingTargets - trainingOutputs;
            this->layers[this->layers.size()-1].outputs = trainingOutputs;
            
            for(unsigned long int k=this->layers.size()-1; k>0 ; k--){
                this->layers[k].gradient = this->layers[k].outputs;
                this->layers[k].gradient.dSigmoid();
                this->layers[k].gradient = this->layers[k].gradient ->* this->layers[k].outputError;
                this->layers[k].gradient = this->layers[k].gradient * this->learningRate;
                this->layers[k].setInputsTransposed();
                Matrix delt(this->layers[k].gradient * this->layers[k].inputsT);
                this->layers[k].setWeightsDelta(delt);
                this->layers[k].weights = this->layers[k].weights + this->layers[k].weightsDelta;
                this->layers[k].biases = this->layers[k].biases + this->layers[k].gradient;
                this->layers[k].setWeightsTransposed();
                this->layers[k].inputError = this->layers[k].weightsT * this->layers[k].outputError;
                this->layers[k-1].outputError = this->layers[k].inputError;
                this->layers[k-1].outputs = this->layers[k].inputs;
                
                this->layers[k].roundToThousandths();
            }
        }
        
    }
}

std::vector<std::vector<double>> NeuralNetwork::predict(std::vector<std::vector<double>> inputData){
    if(inputData[0].size() > this->numNodes.front()){
        std::cout<<"ERROR: Too many input nodes"<<std::endl;
        exit(1);
    } else if (inputData[0].size() < this->numNodes.front()){
        std::cout<<"ERROR: Too few input nodes"<<std::endl;
        exit(1);
    } else {
        Matrix finalOutput(this->feedForward(inputData));
        finalOutput.print();
    }
    return inputData;
}

int NeuralNetwork::checkNumNodesInFile(){
    std::ifstream savedFile;
    savedFile.open(fileName);
    if(savedFile.is_open()){
        int matrixInfo[4];
        std::string scannedValue;
        std::vector<int> numNodes;
        while(!savedFile.eof()){
            std::getline(savedFile, scannedValue);
            if(scannedValue == "~"){
                for(int j=0;j<4;j++){
                    std::getline(savedFile, scannedValue, ' ');
                    matrixInfo[j] = std::stoi(scannedValue);
                }
                if(matrixInfo[1]==0){
                    if(matrixInfo[0] == 0){
                        numNodes.push_back(matrixInfo[3]);
                    }
                    numNodes.push_back(matrixInfo[2]);
                }
            }
        }
        savedFile.close();
        if(this->numNodes == numNodes){
            return 1;
        }else{
            return 0;
        }
    } else {
        return 2;
    }
}

void NeuralNetwork::saveWeightsFile(std::string fileName){
    std::ofstream saveFile;
    saveFile.open(fileName);
    for(int i=0;i<(this->layers.size());i++){
        int numRows = (int)this->layers[i].weights.getNumRows();
        int numCols = (int)this->layers[i].weights.getNumCols();
        saveFile << "~\n";
        saveFile << i << " " << 0 <<" "<< numRows <<" "<< numCols <<" \n";
        for(int y=0;y<numRows;y++){
            //saveFile<<"[";
            for(int x=0;x<(numCols-1);x++){
                saveFile << this->layers[i].weights.getData(y, x)<<" ";
            }
            saveFile << this->layers[i].weights.getData(y, numCols-1)<<" \n";///<<std::endl;
        }
        saveFile << "~\n";
        numRows = (int)this->layers[i].biases.getNumRows();
        numCols = (int)this->layers[i].biases.getNumCols();
        saveFile << i << " " << 1 <<" "<< numRows <<" "<< numCols <<" \n";
        for(int y=0;y<numRows;y++){
            //saveFile<<"[";
            for(int x=0;x<(numCols-1);x++){
                saveFile << this->layers[i].biases.getData(y, x)<<" ";
            }
            saveFile << this->layers[i].biases.getData(y, numCols-1)<<" \n";///<<std::endl;
        }
        saveFile << "\n";
    }
    saveFile.close();
}

void NeuralNetwork::saveWeightsFile(){this->saveWeightsFile(this->fileName);}


void NeuralNetwork::getWeightsFile(std::string fileName){
    int numNodesInFile = this->checkNumNodesInFile();
    if(numNodesInFile == 0){
        std::cout<<"WARNING: File data dimentions do not match this Neural Network."<<std::endl;
        std::cout<<"Would you like to override your current network? (y:n)"<<std::endl;
        char ans;
        std::cin >> ans;
        if(ans=='y' || ans=='Y'){
            this->layers.clear();
            this->numNodes.clear();
            this->setupNetworkFromFile();
        }
    }
    std::ifstream savedFile;
    savedFile.open(fileName);
    if(savedFile.is_open()){
        int matrixInfo[4];
        std::string scannedValue;
        while(!savedFile.eof()){
            std::getline(savedFile, scannedValue);
            if(scannedValue == "~"){
                for(int j=0;j<4;j++){
                    std::getline(savedFile, scannedValue, ' ');
                    matrixInfo[j] = std::stoi(scannedValue);
                }
                for(int y=0;y<matrixInfo[2];y++){
                    for(int x=0;x<matrixInfo[3];x++){
                        std::getline(savedFile, scannedValue, ' ' );
                        if(matrixInfo[1]==0) {
                            this->layers[matrixInfo[0]].weights.setData(y, x, std::stod(scannedValue));
                        } else if(matrixInfo[1]==1) {
                            this->layers[matrixInfo[0]].biases.setData(y, x, std::stod(scannedValue));
                        }
                    }
                }
            }
        }
    } else {
        std::cout<<"ERROR: Could not open file. Error will be ignored, and file will be created later."<<std::endl;
    }
}

void NeuralNetwork::getWeightsFile(){this->getWeightsFile(this->fileName);}

File: Processing/toyNN/NeuralNetwork.py
from Layer import *

class NeuralNetwork:
    def __init__(self, numNodes):
        self.numNodes = numNodes
        self.layers = []
        for i in range(len(numNodes)-1):
            self.layers.append(Layer(numNodes[i],numNodes[i+1]))
        # self.resetWeights();
        self.learningRate = 0.2;
        self.trainingRuns = 1000;
        self.setActivationFunction(Sigmoid)
    def resetWeights(self):
        for i in range(len(self.numNodes)-1):
            self.layers[i].randomize(-1, 1);
    def setActivationFunction(self, func = Sigmoid):
        self.activationFunction = func;
        for i in range(len(self.layers)):
            self.layers[i].setActivationFunction(Sigmoid);
    def toFile(self, fileName):
        fileString = "";
        # self.layers[0].weights.Print()
        # self.layers[0].biases.Print()
        # self.layers[1].weights.Print()
        # self.layers[1].biases.Print()
        for i in range(len(self.layers)):
            numRows = self.layers[i].weights.rows;
            numCols = self.layers[i].weights.cols;
            fileString = fileString + "~\n" + str(i) + " 0 " + str(numRows) + " " + str(numCols) + "\n";
            for y in range(numRows):
                for x in range(numCols-1):
                    fileString = fileString + str(self.layers[i].weights.data[y][x]) + " ";
                fileString = fileString + str(self.layers[i].weights.data[y][numCols-1]) + "\n";
            numRows = self.layers[i].biases.rows;
            numCols = self.layers[i].biases.cols;
            fileString = fileString + "~\n" + str(i) + " 1 " + str(numRows) + " " + str(numCols) + "\n";
            for y in range(numRows):
                for x in range(numCols-1):
                    fileString = fileString + str(self.layers[i].biases.data[y][x]) + " ";
                fileString = fileString + str(self.layers[i].biases.data[y][numCols-1]) + "\n";
        saveBytes(fileName, fileString);
                                    
    def feedForward(self, inputData):
        inputs = fromArray(inputData);
        self.layers[0].setInputs(inputs, False);
        for i in range(1,len(self.layers)):
            self.layers[i].setInputs(self.layers[i-1].solveForOutputs(), False);
        return (self.layers[len(self.layers)-1].solveForOutputs());
    
    def predict(self, inputArray):
        return self.feedForward(inputArray).toArray();
    
    def train(self, inputData, targetData):
        if((len(inputData[0]) != self.numNodes[0]) or (len(targetData[0]) != self.numNodes[-1])):
            print("ERROR: Training data dimentions do not match network dimentions.");
        else:
            for i in range(self.trainingRuns): #Repeat until model has ran for desired number of training runs
                x = floor(random(0,len(inputData)-1));
                trainingOutputs = self.feedForward(inputData[x]); #Matrix of outputs from the training run
                trainingTargets = fromArray(targetData[x]);  #Matrix of targets from the training run
                trainingTargets = ~trainingTargets;
                
                self.layers[len(self.layers)-1].outputError = trainingTargets - trainingOutputs;
                self.layers[len(self.layers)-1].outputs = trainingOutputs;
                
                for k in range((len(self.layers)-1),0,-1):
                    self.layers[k].gradient = (((Map(self.layers[k].outputs, self.activationFunction.dfunc)) ** self.layers[k].outputError) * self.learningRate); #Caculate gradient
                    self.layers[k].weightsDelta = self.layers[k].gradient * (~self.layers[k].inputs);
                    self.layers[k].weights = self.layers[k].weights + self.layers[k].weightsDelta;
                    self.layers[k].biases = self.layers[k].biases + self.layers[k].gradient;
                    self.layers[k].inputError = (~self.layers[k].weights) * self.layers[k].outputError;
                    self.layers[k-1].outputError = self.layers[k].inputError;
                    self.layers[k-1].outputs = self.layers[k].inputs;
    
    def fromFile(self, fileName): 
        lines = loadStrings(fileName);
        matrixInfo = [0,0,0,0]
        newMatrix = False
        i = 0
        while i < len(lines):
            pieces = lines[i].split(' ');
            if (pieces[0] == "~"):
                i += 1;
                pieces = lines[i].split(' ');
                for j in range(len(pieces)):
                    matrixInfo[j] = int(pieces[j]);
                i += 1;
            else:
                for y in range(matrixInfo[2]):
                    pieces = lines[i].split(' ');
                    for x in range(matrixInfo[3]):
                        if(matrixInfo[1]==0):
                            self.layers[matrixInfo[0]].weights.data[y][x] = float(pieces[x]);
                        elif(matrixInfo[1]==1):
                            self.layers[matrixInfo[0]].biases.data[y][x] = float(pieces[x]);
                    i += 1;
        # self.layers[0].weights.Print()
        # self.layers[0].biases.Print()
        # self.layers[1].weights.Print()
        # self.layers[1].biases.Print()
                    

        
